Add S3-compatible cache middleware and cache_page decorator

ERosendo opened the pull request:
This PR introduces full support for caching Django views in S3 using `S3ExpressCacheBackend`. It includes:

### Features

- `CacheMiddlewareS3Compatible`:
     
    A drop-in replacement for Django's CacheMiddleware that:

     - Generates S3-compatible cache keys by placing the key prefix at the start and replacing slashes (/) with dots (.).

     - Automatically uses S3-compatible key generation when the backend is S3ExpressCacheBackend.

     - Falls back to standard Django cache key generation for non-S3 backends.

- `cache_page` decorator:
    A decorator that wraps Django views with `CacheMiddlewareS3Compatible`, providing the same interface as Django’s built-in cache_page but with S3 compatibility.

### Changes

- Refactored cache key generation logic to support S3-safe key names.

- Middleware automatically selects the correct key generation functions depending on the cache backend.

- Added unit tests for:
   - Middleware key generation, cache hits, and default backend behavior.
   - `cache_page` decorator behavior and interaction with S3 backend.
   
- Updates changelog to document new middleware and decorator.

ERosendo committed:
feat(cache): Adds S3-compatible CacheMiddleware

- Introduces `CacheMiddlewareS3Compatible` as a drop-in replacement for Django’s CacheMiddleware
- Adds custom cache key generation to support `S3ExpressCacheBackend`
- Ensures key prefixes appear first and replaces / with . to avoid nested S3 folder structures
- Implements S3-compatible versions of get_cache_key and learn_cache_key
- Falls back to Django’s default behavior when not using the S3 backend

ERosendo committed:
feat(cache): add S3-compatible cache_page decorator

ERosendo committed:
test(cache): Adds tests for S3-compatible cache middleware and decorator

- MiddlewareTest:
  * Verify S3-compatible cache key generation (prefix at start, no slashes)
  * Ensure `CacheMiddlewareS3Compatible` hits the S3 backend
  * Confirm default backend is used when S3 backend is not configured

- CachePageDecoratorTest:
  * Ensure `cache_page` decorator wraps views correctly
  * Verify decorator triggers cache `get` and `set` calls

ERosendo committed:
feat(docs): Updates the changes.md file

ERosendo committed:
chore(tests): Update command to run tests

mlissner reviewed (approved):
I don't have strong opinions on the code, except that there's a surprising amount of funny stuff at a glance. I think that's mostly the nature of the beast, but I'll let Alberto chime in too. 

Two other thoughts:
1. We need an update to the readme to explain these two new features.
2. Let's be sure we use underscores in front of any method or function that we don't want to be public-facing.

Thanks!

albertisfu commented on django_s3_express_cache/middleware.py:
```diff
@@ -0,0 +1,254 @@
+    url_hash = md5(
+        request.build_absolute_uri().encode("ascii"), usedforsecurity=False
+    )
+    raw_key = f"{key_prefix}:views.decorators.cache.cache_page.{method}.{url_hash.hexdigest()}.{ctx.hexdigest()}"
```
In the original cache middleware it generates keys like:

`views.decorators.cache.cache_header.coverage.3ab750a7d3417ea82029eed698fe8145.America/Los_Angeles`

`views.decorators.cache.cache_page.coverage.GET.3ab750a7d3417ea82029eed698fe8145.d41d8cd98f00b204e9800998ecf8427e.America/Los_Angeles`


Where coverage is the key_prefix passed like:

`@cache_page(7 * 60 * 60 * 24, key_prefix="coverage")`

Now, when using the decorator like this:

`@cache_page(7 * 60 * 60 * 24, cache="s3", key_prefix="1-days")`

the `key_prefix` is now being used as the `key_time_prefix`, which becomes the bucket where the cached object is stored.

This means we're missing the original `key_prefix`, which can be confusing at first glance for some users, and I also think it’s useful for debugging purposes.

So I suggest, if possible, keeping the original `key_prefix` parameter passed by the user and introducing a new one, `key_time_prefix`, for the bucket.

So we end up with something like:

`@cache_page(7 * 60 * 60 * 24,  cache="s3", key_prefix="coverage", key_time_prefix ="1-days")`

ERosendo replied:
Thanks for the suggestion. The `CacheMiddlewareS3Compatible` class now accepts a `time_base_prefix` and preserves the `key_prefix`.

albertisfu commented on django_s3_express_cache/middleware.py:
```diff
@@ -0,0 +1,254 @@
+    """
+
+    def _use_s3_backend(self) -> bool:
+        return isinstance(self.cache, S3ExpressCacheBackend)
```
We're calling this on every request. I think the overhead of calling it is minimal, but it might be worth caching its value.

For example, in `__init__`:

self._is_s3_backend = isinstance(self.cache, S3ExpressCacheBackend)


Then, whenever we need it, we can just use the boolean value: `self._is_s3_backend`

ERosendo replied:
Thanks for this suggestion. I've updated the `__init__` method to now trigger the instance check within it

albertisfu commented on django_s3_express_cache/middleware.py:
```diff
@@ -0,0 +1,254 @@
+from django_s3_express_cache import S3ExpressCacheBackend
+
+
+def _generate_cache_key_s3_compatible(request, method, headerlist, key_prefix):
```
Can we add type hints here and in the other methods, please?

ERosendo replied:
I've added type hints to all the helper methods i'm adding in this PR

albertisfu commented on django_s3_express_cache/middleware.py:
```diff
@@ -0,0 +1,254 @@
+        # Use the appropriate cache key generator based on backend:
+        #   • S3 backend -> use S3-compatible helpers.
+        #   • default backend -> use Django's standard helpers.
+        key_func = self._get_cache_key_func
```
Yeah I think we should keep `key_prefix` as suggested in a previous comment.

When I used this middleware with a Redis backend I ended up with this kind of key:

`views.decorators.cache.cache_page..GET.3ab750a7d3417ea82029eed698fe8145.d41d8cd98f00b204e9800998ecf8427e.America/Los_Angeles`

Where the prefix is missing.

Also wouldn't it be better to do:

```
if not self._is_s3_backend:
   return super().process_request(request)
```

At the beginning of this method and later we can just do:

```
cache_key = get_cache_key_s3_compatible(
                request, self.key_prefix, "HEAD", cache=self.cache
            )
```

So we continue relying on the original `process_request()` code for non-S3 backend requests and only use our own version for S3?

If so, we can do something similar for `process_response()`.

ERosendo replied:
I like the idea of using Django’s default implementation when we’re not using the S3 backend. it simplifies the custom class a bit.

albertisfu reviewed (changes requested):
Thanks @ERosendo this looks great and it's working as expected.

I just have some minor suggestions.

ERosendo committed:
feat(middleware): Move backend instance check to  __init__

ERosendo committed:
feat(cache): Add time_based_prefix support to S3 cache middleware

ERosendo committed:
feat(middleware): Enforces time_based_prefix for S3 cache backend

ERosendo committed:
refactor(middleware): Tweaks helpers' signature to take time based prefix

ERosendo committed:
refactor(middleware): Simplifies process_request method

ERosendo committed:
refactor(middleware): Simplifies process_response method

ERosendo committed:
feat(decorator): Add time_based_prefix to cache_page decorator

ERosendo committed:
feat(middleware): Include time_based_prefix in get_cache_key_s3_compatible

ERosendo committed:
fix(tests): Updates tests to use time_base_prefix

ERosendo committed:
docs(readme): Adds section explaining S3-compatible cache_page decorator

Documents how the custom `cache_page` decorator computes time-based prefixes, falls back to Django defaults for non-S3 backends, and can be used as a drop-in replacement for `django.views.decorators.cache.cache_page`

ERosendo commented:
@albertisfu Thanks for your feedback. Here’s a summary of the changes I implemented in response:

- Added type hints to all helper functions introduced in this PR.

- Moved the S3 backend instance check (`isinstance(self.cache, S3ExpressCacheBackend)`) into the middleware’s `__init__`, avoiding repeated checks on every request.

- Introduced a separate time_base_prefix parameter to the CacheMiddlewareS3Compatible class so the time-based prefix used for S3 object keys does not override the regular key_prefix. This also includes:
   - Validations to enforce `time_base_prefix` when using the S3 cache backend.
   - Updated function signatures to explicitly accept both key_prefix and time_base_prefix.

- Cleaned up process_request and process_response so non-S3 backends fall back to Django’s default cache behavior.

- Updated the `cache_page` decorator to compute and pass a `time_base_prefix` (based on the timeout) into the `CacheMiddlewareS3Compatible` class, keeping behavior consistent with the middleware while preserving a drop-in replacement experience for developers.

albertisfu commented on django_s3_express_cache/middleware.py:
```diff
@@ -0,0 +1,270 @@
+        """Store the response in cache when appropriate."""
+        # If not using S3 backend, fall back to default behavior.
+        if not self._is_s3_backend:
+            return super().process_request(request)
```
When using the `django_s3_express_cache.decorators.cache_page decorator` in compatible (Redis backend) mode, it raises the following error:

> The view asgiref.sync.SyncToAsync.__call__ didn't return an HttpResponse object. It returned None instead.

I believe the problem is caused by this line, which may be adjusted as follows:

```suggestion
            return super().process_response(request, response)
```

ERosendo replied:
Thanks for catching this @albertisfu

albertisfu reviewed (changes requested):
Thanks for the changes, @ERosendo. This is almost ready to be merged. I just have one suggestion based on an error I encountered when using the decorator in compatible mode.

ERosendo committed:
fix(middleware): call the correct super method in process_response

Co-authored-by: Alberto Islas <albertisfu@gmail.com>

albertisfu reviewed (approved):
Thanks @ERosendo this is now ready to be merged.
